akka-http/src/main/scala/akka/http/scaladsl/server/directives/MdcLoggingDirectives.scala

```diff
@@ -0,0 +1,63 @@
+package akka.http.scaladsl.server.directives
+
+import akka.event.{
+  DefaultLoggingFilter,
+  DiagnosticMarkerBusLoggingAdapter,
+  LogSource,
+  MarkerLoggingAdapter
+}
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server._
+
+/**
+ * @groupname mdc Mdc logging directives
+ * @groupprio mdc 240
+ */
+trait MdcLoggingDirectives {
+
+  def withMdcLogging: Directive0 =
+    extractActorSystem
+      .flatMap { sys =>
+        mapRequestContext { ctx =>
+          ctx.log match {
+            // If it's already a DiagnosticMarkerBusLoggingAdapter, this is a no-op.
+            case _: DiagnosticMarkerBusLoggingAdapter => ctx
+            // Otherwise, we need to give the context a DiagnosticLoggingAdapter.
+            case _ =>
+              val (str, cls) = LogSource.fromAnyRef(sys, sys)
```

Comment:
LL 27 to 35 were purely type tetris on my part. They work, but I'm not 100% sure I setup the logger the right way. Very open to input here.